### PR TITLE
refactor of packer build

### PR DESCRIPTION
**Pull request by m-juckes-pins**
# Pull Request Template

## Type of change 🧩

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update
- [ ] Other (please explain in the description section above)

## Checklist before requesting a review

- [ ] My code follows the style guidelines of this project
- [ ] My code changes do not include any hardcoded secrets or passwords
- [ ] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] My Tech Lead is aware of any infrastructure changes that will cost money

### CODE REVIEW

File: packer/azure-agents/tools.sh [hunk 1]
```diff
@@ -1,117 +1,137 @@
...
+echo "Installing Checkov..."
+sudo -H python3 -m pip install -U checkov=="${CHECKOV_VERSION}"
 
 # TFLint
+echo "Installing TFLint..."
 curl -s https://raw.githubusercontent.com/terraform-linters/tflint/master/install_linux.sh | bash
 
 # Node / NVM
+echo "Installing Node.js and NVM..."
 curl -sL https://deb.nodesource.com/setup_20.x | sudo -E bash -
```
Comment (m-juckes-pins) [inline packer/azure-agents/tools.sh:97]: look at this line to re-edit

File: packer/azure-agents/config.sh [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+  "main"
+  "restricted"
+  "universe"
+  "multiverse"
+)
+
+COMMON_PACKAGES=(
+  "build-essential"
+  "jq"
+  "libgbm-dev"
```
Comment (m-juckes-pins) [inline packer/azure-agents/config.sh:12]: All these lib packages could be removed potentially, unless a pipeline somewhere requires them?

File: packer/azure-agents/tools.sh [hunk 3]
```diff
@@ -1,117 +1,137 @@
...
-
-# Python
-sudo apt-get install -y --no-install-recommends \
-  python3.7 \
-  python3.7-distutils \
-  python3-pip
+
+# Source the config file
+CONFIG_FILE="/tmp/config.sh"
+if [ -f "$CONFIG_FILE" ]; then
```
Comment (m-juckes-pins) [inline packer/azure-agents/tools.sh:14]: If you wanted to make the script short could remove the if block

File: packer/azure-agents/ubuntu-20-azure-packages.txt
Comment (m-juckes-pins) [inline packer/azure-agents/ubuntu-20-azure-packages.txt:1]: probably best to remove this file and keep it locally or within confluence?

File: infrastructure/provider.tf [hunk 4]
```diff
@@ -1,8 +1,13 @@
 terraform {
+  required_version = ">= 1.7.3, <= 1.9.0"
   required_providers {
     azurerm = {
       source  = "hashicorp/azurerm"
-      version = "3.91.0"
+      version = ">= 1.32.0, 3.110.0"
```
Comment (beejjacobs) [inline infrastructure/provider.tf:6]: Is ">= 1.32.0," needed?

File: packer/azure-agents/config.sh [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+  "libxss1"
+  "libxtst6"
+  "unzip"
+  "xvfb"
+  "python3-pip"
+)
+
+DOCKER_COMPOSE_VERSION="1.29.2"
+
+TERRAFORM_VERSIONS=("1.7.3" "1.9.0")
```
Comment (beejjacobs) [inline packer/azure-agents/config.sh:26]: Should we add a `DEFAULT_TERRAFORM_VERSION` similar to Node?

File: packer/azure-agents/tools.sh [hunk 6]
```diff
@@ -1,117 +1,126 @@
...
-sudo curl -L "https://github.com/docker/compose/releases/download/1.29.2/docker-compose-$(uname -s)-$(uname -m)" -o /usr/local/bin/docker-compose
+sudo curl -L "https://github.com/docker/compose/releases/download/${DOCKER_COMPOSE_VERSION}/docker-compose-$(uname -s)-$(uname -m)" -o /usr/local/bin/docker-compose
 sudo chmod +x /usr/local/bin/docker-compose
 
-# Terraform 1.7.3
-curl -fsSL https://apt.releases.hashicorp.com/gpg | apt-key add -
-sudo apt-add-repository "deb [arch=amd64] https://apt.releases.hashicorp.com $(lsb_release -cs) main"
-sudo apt-get install -y terraform=1.7.3-1
+# Install tfenv
+git clone https://github.com/tfutils/tfenv.git ~/.tfenv
```
Comment (beejjacobs) [inline packer/azure-agents/tools.sh:68]: Should we clone a specific tag rather than just always latest?

File: packer/azure-agents/tools.sh [hunk 7]
```diff
@@ -1,117 +1,126 @@
...
-python3.7 -m pip install -U checkov==2.2.94
+# Checkov via pip
+sudo -H python3 -m pip install -U checkov=="${CHECKOV_VERSION}"
 
 # TFLint
 curl -s https://raw.githubusercontent.com/terraform-linters/tflint/master/install_linux.sh | bash
 
 # Node / NVM
 curl -sL https://deb.nodesource.com/setup_20.x | sudo -E bash -
+sudo apt-get install -y nodejs
```
Comment (beejjacobs) [inline packer/azure-agents/tools.sh:89]: I don't think we need this or line above, as we use `nvm` to install Node?

File: packer/azure-agents/tools.sh [hunk 8]
```diff
@@ -1,117 +1,126 @@
...
 
-sudo apt-get install nodejs
+NVM_DIR="/usr/local/nvm"
+sudo mkdir -p "$NVM_DIR" && sudo chmod -R 777 "$NVM_DIR"
+curl -o- https://raw.githubusercontent.com/creationix/nvm/master/install.sh | NVM_DIR="$NVM_DIR" bash
```
Comment (beejjacobs) [inline packer/azure-agents/tools.sh:93]: This has moved now (it auto redirects), so we should update it I reckon. Also specific a version (could be added to config).

From the [readme](https://github.com/nvm-sh/nvm?tab=readme-ov-file#install--update-script), e.g.
`curl -o- https://raw.githubusercontent.com/nvm-sh/nvm/v0.39.7/install.sh`

File: packer/azure-agents/config.sh (hunk 2 above)
Comment (beejjacobs) [reply packer/azure-agents/config.sh:12]: Yeah I can't think why we'd need any specific libraries with the tooling we're using... might have to try it out with a bit of trial and error!

**Review by beejjacobs — COMMENTED** [review]
Few comments to look at. The approach is great, and the config file makes it much easier to see and update versions!

File: packer/azure-agents/tools.sh (hunk 6 above)
Comment (m-juckes-pins) [reply packer/azure-agents/tools.sh:68]: tfenv hasn't been updated for 2 years, I think it should be ok to keep as is as doubt there will be a major update..?

File: packer/azure-agents/config.sh (hunk 5 above)
Comment (beejjacobs) [reply packer/azure-agents/config.sh:21]: I think good to have a specific version if we don't want to default to latest (e.g. if we update agents before pipelines are ready).

File: packer/azure-agents/config.sh [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+DEFAULT_TERRAFORM_VERSION="1.9.0"
+
+TERRAGRUNT_VERSION="0.55.1"
+
+CHECKOV_VERSION="2.2.94"
+
+NODE_VERSIONS=("20" "18" "16" "15" "14")
+DEFAULT_NODE_VERSION="16"
```
Comment (beejjacobs) [inline packer/azure-agents/config.sh:29]: Can we bump this to 18 please? 16 is out of support now

I don't actually think we need 14,15,16 now actually. None of Core Services are using that.

**Review by beejjacobs — APPROVED** [review]

**Pull request merged**